Add PackageDependenciesDesignTime to cache assets file

ocallesp opened the pull request:
Fixes https://github.com/dotnet/sdk/issues/27738

Implements reading the PackageDefinitions and `PackageDependencies` items in `ResolvePackageAssets`, so that `ResolvePackageDependencies` doesn't need to be called at all unless `EmitLegacyAssetsFileItems` is true. This improve incremental build perf and also significantly improve the design time build perf (23%) for this large solution scenario by saving to the cache file the same items generated from `PreprocessPackageDependenciesDesignTime`.

----------------
Note:
This implementation was taken from `PreprocessPackageDependenciesDesignTime.ExecuteCore`

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAResolvePackageDependenciesTask.cs:
```diff
@@ -40,36 +40,16 @@ public static IEnumerable<object[]> ItemCounts
-                        "simple.dependencies",
-                        new int[] { 113, 2613, 1, 878, 94 },
-                        true
+                        new int[] { 110, 2536, 1, 845, 75 },
```
Did you figure out why some of these numbers are changing?

ocallesp replied:
This unit-test was already broken in release/7.0.2xx. This PR is not affecting the results of this tests.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -160,6 +162,33 @@ public sealed class ResolvePackageAssets : TaskBase
+        /// Items with metadata "ParentTarget" and "ParentPackage", which allows determining the hierarchy of package references.
+        /// </summary>
+        [Required]
+        public ITaskItem[] InputPackageDependencies { get; set; }
+
+        /// <summary>
+        /// Information about each package in the project, with metadata:
+        /// - Name = "MetadataExtractor"
+        /// - Path = "metadataextractor/1.0.0"
+        /// - ResolvedPath = "C:\Users\drnoakes\.nuget\packages\metadataextractor\1.0.0"
+        /// - Type = "package"
+        /// - Version = "2.3.0"
+        /// - DiagnosticLevel = ""
+        /// </summary>
+        [Required]
+        public ITaskItem[] PackageDefinitions { get; set; }
```
`InputPackageDependencies` and `PackageDefinitions` shouldn't be inputs to this task, they should be read from the .assets file as part of this task.

Since they're required inputs and not currently being passed in, I would expect all builds with this change to fail.

ocallesp replied:
The code changed to read InputPackageDependencies and PackageDefinitions from assets file.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -653,7 +679,10 @@ private ITaskItem ReadItem()
         internal sealed class CacheWriter : IDisposable
         {
             private const int InitialStringTableCapacity = 32;
+            private const string ResolvedMetadata = "Resolved";
```
I'd suggest adding this to `MetadataKeys` to match the other metadata names.

ocallesp replied:
changed

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1450,258 @@ private void WritePackageDependencies()
+
+                var allowItemSpecs = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+
+                foreach (var dependency in _packageDependencies)
+                {
+                    if (dependency.HasMetadataValue(MetadataKeys.ParentPackage))
+                    {
+                        // ignore non-top-level packages (those with ParentPackage)
+                        continue;
+                    }
+
+                    var target = dependency.GetMetadata(MetadataKeys.ParentTarget);
+
+                    if (!StringComparer.OrdinalIgnoreCase.Equals(target, _task.TargetFramework))
+                    {
+                        // skip dependencies for other targets
+                        continue;
+                    }
+
+                    allowItemSpecs.Add(dependency.ItemSpec);
+                }
```
This loop should become unnecessary with other changes suggested. We should create a list of dependencies that doesn't have any of the ones that would be filtered out here.

ocallesp replied:
done

ocallesp replied:
removed in the code cleanup

dsplaisted reviewed (commented):
See comments for ways we should be able to eliminate a bunch of the code here and process less data.

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -726,6 +756,24 @@ public CacheWriter(ResolvePackageAssets task)
+                    }
+                });
+
+                void ReadProjectFileDependencies(string frameworkAlias)
```
changed.

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -726,6 +756,24 @@ public CacheWriter(ResolvePackageAssets task)
                     ComputePackageExclusions();
                 }
+
+                Dictionary<string, string> CreateTargetNameToAlisMap() => _lockFile.Targets.ToDictionary(t => t.Name, t =>
```
changed

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -309,6 +308,7 @@ Copyright (c) .NET Foundation. All rights reserved.
       <Output TaskParameter="TransitiveProjectReferences" ItemName="_TransitiveProjectReferences" />
       <Output TaskParameter="PackageFolders" ItemName="AssetsFilePackageFolder" />
       <Output TaskParameter="PackageDependencies" ItemName="PackageDependencies" />
+      <Output TaskParameter="PackageDependenciesDesignTime" ItemName="_PackageDependenciesDesignTime" />
```
If `PreprocessPackageDependenciesDesignTime` is deleted, then `ResolvePackageDependenciesDesignTime` will be empty. Should we keep empty Targets ? it seems that the dependency is not needed.

ocallesp replied:
done

drewnoakes replied:
@ocalles while the `ResolvePackageDependenciesDesignTime` target no longer has any tasks within it, it still specifies `Returns="@(_PackageDependenciesDesignTime)"`, meaning those items are passed back to callers (such as the .NET Project System).

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1450,258 @@ private void WritePackageDependencies()
+                    {
+                        // raise each library in the target
+                        GetPackageDependenciesPerTarget(target);
+                    }
```
fixed

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -726,6 +756,24 @@ public CacheWriter(ResolvePackageAssets task)
+
+                void ReadProjectFileDependencies(string frameworkAlias)
+                {
+                    _projectFileDependencies = _lockFile.GetProjectFileDependencySet(frameworkAlias);
```
fixed

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1450,258 @@ private void WritePackageDependencies()
+                        }
+
+                        // get sub package dependencies
+                        GetDependencies(package, target.Name, resolvedPackageVersions, transitiveProjectRefs);
```
That makes sense.
Removed.

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1450,258 @@ private void WritePackageDependencies()
+                GetPackageDependencies();
+
+                List<ITaskItem> _packageDefinitions = new List<ITaskItem>();
+                GetPackageDefinitions();
```
done

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -344,16 +344,7 @@ Copyright (c) .NET Foundation. All rights reserved.
-
-      <Output TaskParameter="PackageDependenciesDesignTime" ItemName="_PackageDependenciesDesignTime" />
-    </PreprocessPackageDependenciesDesignTime>
+          DependsOnTargets="ResolvePackageAssets;RunResolvePackageDependencies;ResolveAssemblyReferencesDesignTime">
 
   </Target>
```
```suggestion
          DependsOnTargets="ResolvePackageAssets;RunResolvePackageDependencies;ResolveAssemblyReferencesDesignTime" />
```

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1438,203 @@ private void WritePackageDependencies()
 
+            // This implementation was taken from PreprocessPackageDependenciesDesignTime.ExecuteCore
+            // with some minor modifications to use WriteItem instead of saving it into an array.
+            private void WritePackageDependenciesDesignTime()
```
I think we can simplify this method (and the nested methods within it) more.  In `GetPackageDefinitions` we are looping through the lock file libraries and creating a `PackageDefinitionItem` for each one.  Then in the top-level method we are looping through those PackageDefinitionItems, skipping the ones that aren't top-level dependencies, and transforming them into the items that we write out.

Can we try getting rid of the `GetPackageDefinitions` method and the intermediate `PackageDefinitionItem` type, and combine this logic into one loop in the top-level method?

ocallesp replied:
done

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1438,203 @@ private void WritePackageDependencies()
+
+                string ResolvePackagePath(LockFileLibrary package)
+                {
+                    if (package.IsProject())
+                    {
+                        var relativeMSBuildProjectPath = package.MSBuildProject;
+
+                        if (string.IsNullOrEmpty(relativeMSBuildProjectPath))
+                        {
+                            throw new BuildErrorException(Strings.ProjectAssetsConsumedWithoutMSBuildProjectPath, package.Name, _task.ProjectAssetsFile);
+                        }
+
+                        return GetAbsolutePathFromProjectRelativePath(relativeMSBuildProjectPath);
+                    }
+                    else
```
I think we can get rid of all the path logic for projects here, as around line 1475 we are filtering out anything that doesn't have a type of Package or Unresolved.  Given that we can probably just remove the whole `ResolvePackagePath` method and replace it with a call to `_packageResolver.GetPackageDirectory`.

ocallesp replied:
done

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1438,203 @@ private void WritePackageDependencies()
+                    }
+                }
+
+                string GetAbsolutePathFromProjectRelativePath(string path)
```
We should also be able to delete this method.

ocallesp replied:
done

dsplaisted reviewed (commented):
This is looking good, thanks.

Have you made any progress with @drewnoakes on running the project system tests with these changes?

Also do you have any perf testing results?  It would be great to be able to present those (or a demo) at the next team showcase.

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1438,203 @@ private void WritePackageDependencies()
+                        _compileTimeTarget.Libraries
+                            .Where(lib => lib.IsTransitiveProjectReference(_lockFile, ref projectFileDependencies, frameworkAlias))
+                            .Select(pkg => pkg.Name),
+                        StringComparer.OrdinalIgnoreCase);
```
deleted

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -344,16 +344,7 @@ Copyright (c) .NET Foundation. All rights reserved.
           Returns="@(_PackageDependenciesDesignTime)"
-          DependsOnTargets="RunResolvePackageDependencies;ResolveAssemblyReferencesDesignTime">
-
-    <PreprocessPackageDependenciesDesignTime
```
It seems that we cannot remove `PreprocessPackageDependenciesDesignTime`  since it is used by  `CollectSDKReferencesDesignTime`.

https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/CollectSDKReferencesDesignTime.cs#L37

ocallesp replied:
done

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1438,203 @@ private void WritePackageDependencies()
+            // with some minor modifications to use WriteItem instead of saving it into an array.
+            private void WritePackageDependenciesDesignTime()
+            {
+                HashSet<string> projectFileDependencies = _lockFile.GetProjectFileDependencySet(_compileTimeTarget.Name);
```
moved inside the method

ocallesp commented:
> This is looking good, thanks.
> 
> Have you made any progress with @drewnoakes on running the project system tests with these changes?
> 
> Also do you have any perf testing results? It would be great to be able to present those (or a demo) at the next team showcase.

Yes. I am planning to run some performance tests once the pr is ready.

dsplaisted commented:
> > This is looking good, thanks.
> > Have you made any progress with @drewnoakes on running the project system tests with these changes?
> > Also do you have any perf testing results? It would be great to be able to present those (or a demo) at the next team showcase.
> 
> Yes. I am planning to run some performance tests once the pr is ready.

What about the project system tests?  The project system is what consumes data from the `ResolvePackageDependenciesDesignTime` target in order to display the dependency information in the Solution Explorer tree.  Since we are changing that logic and probably don't have good coverage of this in the SDK repo, we should find a way to run the project system tests with these SDK changes.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenThatWeWantToGetDependenciesViaDesignTimeBuild.cs:
```diff
@@ -1,399 +0,0 @@
-
-namespace Microsoft.NET.Build.Tasks.UnitTests
-{
-    public class GivenThatWeWantToGetDependenciesViaDesignTimeBuild
```
We have moved this functionality to the `ResolvePackageAssets` task.  Can we add unit tests similar to these ones that are being deleted to the tests for `ResolvePackageAssets`?

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/CollectSDKReferencesDesignTime.cs:
```diff
@@ -34,15 +34,36 @@ public class CollectSDKReferencesDesignTime : TaskBase
             SDKReferencesDesignTime = sdkDesignTimeList.ToArray();
         }
 
+        internal static HashSet<string> GetImplicitPackageReferences(string defaultImplicitPackages)
```
This method was also copied into `ResolvePackageAssets`.  Can you unify these?

ocallesp replied:
done

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1428,141 @@ private void WritePackageDependencies()
+
+                foreach (var package in _lockFile.Libraries)
+                {
+                    if (string.IsNullOrEmpty(package.Name) || !allowItemSpecs.Contains(package.Name))
```
The `allowedItemSpecs` are of the form `{packageName}/{packageVersion}` (from the `GetPackageId` method).  So `package.Name` will never match this.

I missed when this was introduced when the `PackageDefinitionItem` type was added.

Are you doing any ad-hoc testing of these changes?  I think this means that no packages would be displayed as dependencies in Solution Explorer.

ocallesp replied:
fixed

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1428,141 @@ private void WritePackageDependencies()
+                    if (dependencyType == DependencyType.Package ||
+                        dependencyType == DependencyType.Unresolved)
+                    {
+                        WriteItem(package.Name);
```
I think the itemSpec needs to include the version, in the form returned by `GetPackageId`.

ocallesp replied:
fixed

marcin-krystianc commented:
@ocallesp @dsplaisted 
I've run a benchmark based on https://github.com/dotnet/sdk/issues/27738#issue-1364794138 and the results look very promising:

before (commit 85a710726e869ee09668248bae3aa73126c370d6):
```
Time Elapsed 00:01:27.60
Time Elapsed 00:01:25.84
Time Elapsed 00:01:25.44
```

After (commit db85ef3a5882f78a133ed1181eaf5cd91c4e22fe):
```
Time Elapsed 00:01:17.74
Time Elapsed 00:01:04.85
Time Elapsed 00:01:05.54
```

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1428,121 @@ private void WritePackageDependencies()
+                {
+                    List<string> dependencies = new();
+                    HashSet<string> projectFileDependencies = _lockFile.GetProjectFileDependencySet(_compileTimeTarget.Name);
+
+                    var resolvedPackageVersions = _compileTimeTarget.Libraries
+                        .ToDictionary(pkg => pkg.Name, pkg => pkg.Version.ToNormalizedString(), StringComparer.OrdinalIgnoreCase);
+
+                    string frameworkAlias = _lockFile.GetLockFileTargetAlias(_compileTimeTarget);
```
Appears to be unused and doing work for nothing.

```suggestion
```

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1428,121 @@ private void WritePackageDependencies()
+                    }
+                }
+
+                List<string> GetPackageDependencies()
+                {
+                    List<string> dependencies = new();
```
This can ge a generator function that `yield return`s its values, so we don't have to allocate and collect a `List<string>` and backing `string[]`.

```suggestion
                IEnumerable<string> GetPackageDependencies()
                {
```

dsplaisted replied:
Or it could just return a HashSet since that's what it's transformed into anyway.

ocallesp replied:
replaced with yield return

drewnoakes replied:
Using `HashSet<>` would avoid allocating the enumerator.

ocallesp replied:
fixed

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1428,121 @@ private void WritePackageDependencies()
+                    {
+                        if (projectFileDependencies.Contains(package.Name))
+                        {
+                            string packageId = GetPackageId(package);
+
+                            // itemSpec
+                            dependencies.Add(packageId);
```
nit: Use the variable name to document the value's semantic meaning, instead of a separate comment.

```suggestion
                            string itemSpec = GetPackageId(package);

                            dependencies.Add(itemSpec);
```

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1428,121 @@ private void WritePackageDependencies()
+                {
+                    string target = _task.TargetFramework ?? "";
+
+                    var messages = _lockFile.LogMessages.Where(log => log.LibraryId == package.Name && log.TargetGraphs
+                                    .Select(tg =>
+                                    {
+                                        var parsedTargetGraph = NuGetFramework.Parse(tg);
+                                        var alias = _lockFile.PackageSpec.TargetFrameworks.FirstOrDefault(tf => tf.FrameworkName == parsedTargetGraph)?.TargetAlias;
+                                        return alias ?? tg;
+                                    }).Contains(target));
```
I find this indentation hard to read. For example, it's not clear what the `Contains` is being called on.

Something like this?

```suggestion
                    var messages = _lockFile.LogMessages.Where(log =>
                        log.LibraryId == package.Name &&
                        log.TargetGraphs.Any(tg =>
                        {
                            var parsedTargetGraph = NuGetFramework.Parse(tg);
                            var alias = _lockFile.PackageSpec.TargetFrameworks
                                .FirstOrDefault(tf => tf.FrameworkName == parsedTargetGraph)
                                ?.TargetAlias ?? tg;
                            return alias == target;
                        });
```

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1428,121 @@ private void WritePackageDependencies()
+                                        return alias ?? tg;
+                                    }).Contains(target));
+
+                    if (messages.Count() == 0)
```
This code double-enumerates `messages`. In the case that there are many messages, you don't need to enumerate all of them here:

```suggestion
                    if (!messages.Any())
```

drewnoakes reviewed (commented):
Some comments and suggestions to review before we meet this week @ocallesp.

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenThatWeWantToGetDependenciesViaDesignTimeBuild.cs:
```diff
@@ -12,388 +18,129 @@ public class GivenThatWeWantToGetDependenciesViaDesignTimeBuild
+            `lib/netstandard2.0/Newtonsoft.Json.dll` : {
+                `related` : `.pdb;.xml`
+            }
+        },
```
@dsplaisted what values can be passed to the json file to avoid the null value in `packagePath `from `_packagePathResolver`

https://github.com/dotnet/sdk/blob/main/src/Tasks/Microsoft.NET.Build.Tasks/NuGetPackageResolver.cs#L49

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1428,121 @@ private void WritePackageDependencies()
             }
 
+            // This implementation was taken from PreprocessPackageDependenciesDesignTime.ExecuteCore
+            // with some minor modifications to use WriteItem instead of saving it into an array.
```
removed from code and added to the pr

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.PackageDependencyResolution.targets:
```diff
@@ -196,14 +196,12 @@ Copyright (c) .NET Foundation. All rights reserved.
       TargetFramework="$(TargetFramework)"
-      ContinueOnError="ErrorAndContinue">
+      ContinueOnError="ErrorAndContinue"
+      Condition="'$(EmitLegacyAssetsFileItems)' == 'true'">
```
added

ocallesp commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenThatWeWantToGetDependenciesViaDesignTimeBuild.cs:
```diff
@@ -12,388 +18,129 @@ public class GivenThatWeWantToGetDependenciesViaDesignTimeBuild
-            Assert.True(item.GetBooleanMetadata(MetadataKeys.IsImplicitlyDefined));
+            string projectAssetsJsonPath = Path.GetTempFileName();
+            string projectCacheAssetsJsonPath = Path.GetTempFileName();
+            var assetsContent = @"{
```
changes to use raw string

drewnoakes commented on do.bat:
```diff
@@ -0,0 +1 @@
+build.cmd
```
Did you intend to commit these four `.bat` files?

ocallesp replied:
removed

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolvePackageAssets.cs:
```diff
@@ -1405,6 +1428,121 @@ private void WritePackageDependencies()
+                    var packageVersion = package.Version.ToNormalizedString();
+                    string packageId = $"{package.Name}/{packageVersion}";
+
+                    if (string.IsNullOrEmpty(package.Name) || !allowItemSpecs.Contains(packageId))
+                    {
+                        // Name is required
+                        // We are not interested in this definition (not top-level, or wrong target)
+                        continue;
+                    }
```
Alternatively, change the comment. It's confusing as it currently stands.

```suggestion
                    if (string.IsNullOrEmpty(package.Name) || !allowItemSpecs.Contains(packageId))
                    {
                        // Only include packages from the allow list.
                        // This excludes transitive packages and those from other targets.
                        continue;
                    }
```

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenThatWeWantToGetDependenciesViaDesignTimeBuild.cs:
```diff
@@ -15,87 +15,89 @@ namespace Microsoft.NET.Build.Tasks.UnitTests
     public class GivenThatWeWantToGetDependenciesViaDesignTimeBuild
     {
-        [Fact]
+        [Fact(Skip = "TODO: fix packagePath from _packagePathResolver returns null")]
```
Are you planning to fix this as part of this PR?

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenThatWeWantToGetDependenciesViaDesignTimeBuild.cs:
```diff
@@ -15,87 +15,89 @@ namespace Microsoft.NET.Build.Tasks.UnitTests
-        `compile` : {
-            `lib/netstandard2.0/Newtonsoft.Json.dll` : {
-                `related` : `.pdb;.xml`
+            var assetsContent =
+"""
+{
+  "version" : 3,
+  "targets" : {
```
Raw strings can be indented, and the leading white space is ignored.

```c#
    var string =
        """
        Hello,
        World!
        """;
```

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAResolvePackageDependenciesTask.cs:
```diff
@@ -40,36 +40,16 @@ public static IEnumerable<object[]> ItemCounts
-                        "simple.dependencies",
-                        new int[] { 113, 2613, 1, 878, 94 },
-                        true
+                        new int[] { 110, 2536, 1, 845, 75 },
                     },
                     new object[] {
                         "simple.dependencies",
-                        new int[] { 113, 0, 0, 878, 0 },
-                        false
+                        new int[] { 113, 2613, 1, 877, 96 },
                     },
```
The two cases that were dropped here where `emitLegacyAssetsFileItems` was false, which is expected.

What's unexepected is that the numbers on the remaining cases changed slightly.

ocallesp replied:
@drewnoakes I executed the test in release/7.0.2xx and this unit-test is broken.

ocallesp commented:
It seems that the github build is failing due to some infrastructure issue.

ocalles committed:
Add output PackageDependenciesDesignTime to cache assets file

- ResolvePackageDependencies only when EmitLegacyAssetsFileItem is true
- Removing all the conditions that depend on EmitLegacyAssetsFileItems
since there is case when this variable is false.
- Get PackageDefinitions from Assets fileA dd output PackageDependenciesDesignTime to cache file
- Remove call to task PreprocessPackageDependenciesDesignTime
- Use _compileTimeTarget.Name instead of frameworkAlias
- Delete PreprocessPackageDependenciesDesignTime task entirely
- combine GetPackageDefinitions() with the loop
- Add basic unit-tests for PackageDependenciesDesignTime

marcin-krystianc commented:
> > I've run a benchmark based on [#27738 (comment)](https://github.com/dotnet/sdk/issues/27738#issue-1364794138) and the results look very promising:
> 
> @marcin-krystianc Could you share the setup you used to run this benchmark?

Sure, 
- I'm using AWS machine (`m5a.xlarge` instance) running Ubuntu os: 
- I've used https://github.com/marcin-krystianc/TestSolutions/tree/master/LargeAppWithPrivatePackagesCentralisedNGBVRemoved
- Go to `LargeAppWithPrivatePackagesCentralisedNGBVRemoved\solution` and run `git clean -fxd & dotnet restore`
- Benchmark:
```
seq 5 | xargs -I -- \
dotnet msbuild \
-target:"GetSuggestedWorkloads;_CheckForInvalidConfigurationAndPlatform;ResolveReferences;ResolveProjectReferences;ResolveAssemblyReferences;ResolveComReferences;ResolveNativeReferences;ResolveSdkReferences;ResolveFrameworkReferences;ResolvePackageDependenciesDesignTime;Compile;CoreCompile" \
/p:AndroidPreserveUserData=True \
/p:AndroidUseManagedDesignTimeResourceGenerator=True \
/p:BuildingByReSharper=True \
/p:BuildingProject=False \
/p:BuildProjectReferences=False \
/p:ContinueOnError=ErrorAndContinue \
/p:DesignTimeBuild=True \
/p:DesignTimeSilentResolution=true \
/p:JetBrainsDesignTimeBuild=True \
/p:ProvideCommandLineArgs=True \
/p:ResolveAssemblyReferencesSilent=true \
/p:SkipCompilerExecution=True \
/p:TargetFramework=net5.0 \
/v:n \
/m:1 \
/nr:false \
/flp:v="n;PerformanceSummary" \
/clp:Summary \
/clp:PerformanceSummary \
>> log.txt
```
- Results, before:
```
Time Elapsed 00:01:25.15
Time Elapsed 00:01:22.67
Time Elapsed 00:01:22.24
Time Elapsed 00:01:22.55
Time Elapsed 00:01:22.73
```
- Results, after:
```
Time Elapsed 00:01:16.28
Time Elapsed 00:01:02.96
Time Elapsed 00:01:02.70
Time Elapsed 00:01:02.60
Time Elapsed 00:01:02.81
```

ocalles committed:
Add unit-test for PackageDependenciesDesignTime

ocallesp commented:
Remining work: Implement Unit-tests

drewnoakes commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenThatWeWantToGetDependenciesViaDesignTimeBuild.cs:
```diff
@@ -0,0 +1,839 @@
+        [Fact]
+        public void ItShouldIgnoreAllDependenciesWithTypeNotEqualToPackageOrUnresolved()
+        {
+            string projectAssetsJsonPath = Path.GetTempFileName();
+            string projectCacheAssetsJsonPath = Path.GetTempFileName();
```
The test should delete both of these files when it completes, to avoid polluting the temp dir.

dsplaisted committed:
Fix test paths

ocallesp commented:
@marcin-krystianc  I see that the repository https://github.com/marcin-krystianc/TestSolutions/tree/master/LargeAppWithPrivatePackagesCentralisedNGBVRemoved  does not contain any readme file nor a license.

We would like to know if we are allowed to share this with folks.

marcin-krystianc commented:
> @marcin-krystianc I see that the repository https://github.com/marcin-krystianc/TestSolutions/tree/master/LargeAppWithPrivatePackagesCentralisedNGBVRemoved does not contain any readme file nor a license.
> 
> We would like to know if we are allowed to share this with folks.

Hi @ocallesp, I'll add some readme and license information, but of course it is ok to share it with others.

ocallesp commented:
@marcin-krystianc   I appreciate your willingness to share your repository with others.
